### PR TITLE
Convert more text blocks to JSON format

### PR DESCRIPTION
**Pull request by RoboErikG**
Converts text_charAt, text_indexOf, text_isEmpty, and text_length
to JSON. Relevant changes are in blocks/text.js and msg/messages.js

Includes a rebuild to pick up message changes.

---
This change is [Reviewable](https://reviewable.io/reviews/google/blockly/1163)

### CODE REVIEW

File: blocks/text.js [hunk 1]
```diff
@@ -847,14 +756,122 @@ Blockly.Constants.Text.TEXT_JOIN_EXTENSION = function() {
...
+    this.updateAt_(true);
     // Assign 'this' to a variable for use in the tooltip closure below.
     var thisBlock = this;
     this.setTooltip(function() {
-      return Blockly.Msg.TEXT_APPEND_TOOLTIP.replace('%1',
-          thisBlock.getFieldValue('VAR'));
+      var where = thisBlock.getFieldValue('WHERE');
+      var tooltip = Blockly.Msg.TEXT_CHARAT_TOOLTIP;
+      if (where == 'FROM_START' || where == 'FROM_END') {
+        var msg = (where == 'FROM_START') ?
```
Comment (rachel-fenichel) [inline blocks/text.js]: Here also, check that the message exists.

File: blocks/text.js [hunk 2]
```diff
@@ -847,14 +756,122 @@ Blockly.Constants.Text.TEXT_JOIN_EXTENSION = function() {
 };
 
 Blockly.Constants.Text.TEXT_APPEND_TOOLTIP_EXTENSION = function() {
+  // Assign 'this' to a variable for use in the tooltip closure below.
+  var thisBlock = this;
+  this.setTooltip(function() {
+    return Blockly.Msg.TEXT_APPEND_TOOLTIP.replace('%1',
```
Comment (RoboErikG) [inline blocks/text.js]: done.

File: blocks/text.js [hunk 3]
```diff
@@ -847,14 +756,122 @@ Blockly.Constants.Text.TEXT_JOIN_EXTENSION = function() {
...
+  ],
+
+  /**
+   * Create XML to represent whether there is an 'AT' input.
+   * @return {!Element} XML storage element.
+   * @this Blockly.Block
+   */
+  mutationToDom: function() {
+    var container = document.createElement('mutation');
+    var isAt = this.getInput('AT').type == Blockly.INPUT_VALUE;
```
Comment (RoboErikG) [inline blocks/text.js]: isAt is set to false and the mutation adds the attribute "at"="false"

File: blocks/text.js (hunk 1 above)
Comment (RoboErikG) [inline blocks/text.js]: done.

**Comment by RoboErikG** [comment]
Verbal :lgtm: from Rachel.

---

Review status: 0 of 80 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1163#-:-KmSN8e8VHyI2dpDZhRP:b-9vaecg)*

**Comment by NeilFraser** [comment]
This PR deleted hundreds of translations, resulting in English leaking in other languages.  Is there a reason these translations weren't salvaged?

**Comment by NeilFraser** [comment]
The TEXT_CHARAT_TITLE message in this PR is broken.  For example:
`Blockly.Msg.TEXT_CHARAT_TITLE = "IN TEXT %1 X %2 Y";`
Renders as:
screen shot 2017-10-26 at 15 03 29
Note that 'X' is lost, and Y is in front of %1.  I'm also not sure what the %1 and %2 represent, since if they are reversed the block's inputs are unchanged (though different text is lost).

Related, there's also a Blockly.Msg.TEXT_CHARAT_TAIL

**Comment by RoboErikG** [comment]
This PR added those strings. They didn't previously exist. I think they must have been lost later after being translated.

**Pull request merged**